src/routing_node.rs

```diff
@@ -1138,35 +1154,139 @@ impl RoutingNode {
+                                        }
+                                    }
+                                    Data::ImmutableData(ref immutable_data) => {
+                                        if self.cache_options.immutable_data_caching_enabled() {
+                                            debug!("Caching ImmutableData {:?}", data.name());
+                                            // TODO verify data
+                                            data_cache.insert(data.name(), data.clone());
+                                        }
+                                    }
+                                }
+                            }
+                            _ => {}
                         }
-                    }
-                    _ => None
+
+                    },
+                    _ => {}
                 }
-            }
-            _ => None,
+            },
+            None => {}
         }
+            
     }
 
-    fn handle_cache_put(&mut self, response: ExternalResponse) {
-        match response {
-            ExternalResponse::Get(data, _, _) => {
-                match data {
-                    Data::ImmutableData(ref immutable_data) => {
-                        self.data_cache.insert(immutable_data.name(), data.clone());
-                    }
-                    _ => {}
+    fn handle_cache_get(&mut self, message: &RoutingMessage) -> Option<Content> {
+        match self.data_cache {
+            Some(ref mut data_cache) => {
+                match message.content.clone() {
+                    Content::ExternalRequest(request) => {
+                        match request {
+                            ExternalRequest::Get(data_request, _) => {
+                                match data_request {
+                                    DataRequest::PlainData(data_name) => {
+                                        if self.cache_options.plain_data_caching_enabled() {
+                                            match data_cache.get(&data_name) {
+                                                Some(data) => {
+                                                    let response =
+                                                        ExternalResponse::Get(
+                                                            data.clone(),
+                                                            data_request,
+                                                            None);
+                                                    return Some(Content::ExternalResponse(response));
+                                                },
+                                                None => return None
+                                            }
+                                        }
+                                        return None;
+                                    }
+                                    DataRequest::StructuredData(data_name, version) => {
```

Comment:
It is only a name, but for code clarity it is worthwhile fixing, the `u64` in the `DataRequest::StructuredData(identifier_name, type_tag)` - it is not the version number of the StructuredData (the SAFE network does not maintain the version history of an SD, the internal version number functions only to linearise the history in clients control)